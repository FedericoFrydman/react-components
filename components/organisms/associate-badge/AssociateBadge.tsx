import { forwardRef, ForwardRefRenderFunction, Ref } from 'react';
import cn from 'classnames';

import { Avatar } from '../../molecules/avatar/Avatar';
import { Container } from '../../atoms/container/Container';
import { Button } from '../../molecules/button/Button';
import { Dropdown, DropdownItem } from '../dropdown/Dropdown';
import { Heading } from '../../atoms/heading/Heading';
import { Icon } from '../../atoms/icon/Icon';

import './associate-badge.scss';

export type AssociateBadgeChildrenBackgroundColors = 'white' | 'gray';

export interface AssociateBadgeProps extends Omit<React.HTMLProps<HTMLDivElement>, 'onClick' | 'css'> {
  /**
   * title of the Associate
   */
  title?: string;
  /**
   * Name of the Associate
   */
  name?: string;
  /**
   * Formatted phone number of the Associate
   */
  phone?: string;
  /**
   * URL of the Associate avatar
   */
  avatarUrl?: string;
  /**
   * Character length of the title and name text before truncation. Defaults to 35.
   */
  truncationLength?: number;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
  /**
   * onClick handler for entire Associate Badge
   */
  onBadgeClick?: () => void;
  /**
   * onClick handler for phone button click
   */
  onPhoneClick?: () => void;
  /**
   * onClick handler for message button click
   */
  onMessageClick?: () => void;
}

const AssociateBadgeComponent: ForwardRefRenderFunction<HTMLDivElement, AssociateBadgeProps> = (
  {
    avatarUrl,
    classNames = [],
    name,
    phone,
    title,
    truncationLength = 35,
    onBadgeClick,
    onPhoneClick,
    onMessageClick,
    ...props
  }: AssociateBadgeProps,
  ref: AssociateBadgeProps['ref'],
) => {
  const truncatedTitle = title.length > truncationLength ? title.substring(0, truncationLength) + '...' : title;
  const truncatedName = name.length > truncationLength ? name.substring(0, truncationLength) + '...' : name;

  const classAssociateBadge = 'lsux-associate-badge';

  return (
    <Container
      {...props}
      classNames={[classAssociateBadge, 'px-4', 'py-5', cn(classNames)]}
      ref={ref}
      onClick={onBadgeClick}
    >
      <div>
        <Avatar avatarType={avatarUrl ? 'user-photo' : 'user-icon'} imageURL={avatarUrl} />
      </div>
      <div className="pl-2">
        <Heading as="T16" text={truncatedTitle} classNames={[`${classAssociateBadge}--title`, 'pl-3']} />
        <Dropdown.Root>
          <Dropdown.Trigger>
            <Button
              aria-label="Menu button"
              iconRight="arrow_caret_down_md"
              iconColor="P200"
              iconSize="medium"
              label={truncatedName}
              shape="round"
              variant="tertiary"
            />
          </Dropdown.Trigger>
          <Dropdown.Portal>
            {phone ? (
              <DropdownItem onClick={onPhoneClick}>
                Call {phone}
                <div className="right-slot">
                  <Icon name="comm_phone_default" size="small" />
                </div>
              </DropdownItem>
            ) : null}
            <DropdownItem onClick={onMessageClick}>
              Message Me
              <div className="right-slot">
                <Icon name="comm_email_default" size="small" />
              </div>
            </DropdownItem>
          </Dropdown.Portal>
        </Dropdown.Root>
      </div>
    </Container>
  );
};

export const AssociateBadge = forwardRef<HTMLDivElement, AssociateBadgeProps>(AssociateBadgeComponent);
