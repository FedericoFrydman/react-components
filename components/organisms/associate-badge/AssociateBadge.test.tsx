import { createRef } from 'react';
import { render, screen } from '@testing-library/react';
import { shallow, ShallowWrapper } from 'enzyme';
import cn from 'classnames';

import { AssociateBadge, AssociateBadgeProps } from './AssociateBadge';

describe('Associate Badge Component', () => {
  const testId = 'test-associate-badge';
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(
      <AssociateBadge
        avatarUrl="https://api.legalshield.com/v2/public/associates/avatar/127525095-1596790237.195000.jpg"
        name="Sherry Swedenborg"
        onMessageClick={() => console.log('onMessageClick')}
        onPhoneClick={() => console.log('onPhoneClick')}
        phone="(238) 432-6343"
        title="LegalShield Independent Associate"
        data-testid={testId}
      />,
    );
  });

  test('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the component the provided classNames', () => {
          const badge = renderAssociateBadge({ classNames });

          expect(badge).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLDivElement>();

    it('provides a ref to the associate badge', () => {
      expect(renderAssociateBadge({ ref })).toEqual(ref.current);
    });
  });

  const renderAssociateBadge = (props: AssociateBadgeProps = {}): HTMLElement => {
    render(
      <AssociateBadge
        {...props}
        avatarUrl="https://api.legalshield.com/v2/public/associates/avatar/127525095-1596790237.195000.jpg"
        name="Sherry Swedenborg"
        onMessageClick={() => console.log('onMessageClick')}
        onPhoneClick={() => console.log('onPhoneClick')}
        phone="(238) 432-6343"
        title="LegalShield Independent Associate"
        data-testid={testId}
      />,
    );

    return screen.getByTestId(testId);
  };
});
