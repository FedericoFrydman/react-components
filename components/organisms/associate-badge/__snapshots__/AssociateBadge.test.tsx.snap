// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Associate Badge Component should render the component 1`] = `
<ForwardRef(ContainerComponent)
  classNames={
    Array [
      "lsux-associate-badge",
      "px-4",
      "py-5",
      "",
    ]
  }
  data-testid="test-associate-badge"
>
  <div>
    <ForwardRef(AvatarComponent)
      avatarType="user-photo"
      imageURL="https://api.legalshield.com/v2/public/associates/avatar/127525095-1596790237.195000.jpg"
    />
  </div>
  <div
    className="pl-2"
  >
    <ForwardRef(HeadingComponent)
      as="T16"
      classNames={
        Array [
          "lsux-associate-badge--title",
          "pl-3",
        ]
      }
      text="LegalShield Independent Associate"
    />
    <DropdownRoot>
      <DropdownTrigger>
        <ForwardRef(ButtonComponent)
          aria-label="Menu button"
          iconColor="P200"
          iconRight="arrow_caret_down_md"
          iconSize="medium"
          label="Sherry Swedenborg"
          shape="round"
          variant="tertiary"
        />
      </DropdownTrigger>
      <DropdownPortal>
        <DropdownItem
          onClick={[Function]}
        >
          Call 
          (238) 432-6343
          <div
            className="right-slot"
          >
            <ForwardRef(IconComponent)
              name="comm_phone_default"
              size="small"
            />
          </div>
        </DropdownItem>
        <DropdownItem
          onClick={[Function]}
        >
          Message Me
          <div
            className="right-slot"
          >
            <ForwardRef(IconComponent)
              name="comm_email_default"
              size="small"
            />
          </div>
        </DropdownItem>
      </DropdownPortal>
    </DropdownRoot>
  </div>
</ForwardRef(ContainerComponent)>
`;
